src/search_controller.ts

```diff
@@ -0,0 +1,488 @@
+import { debounce, DebouncedFunc } from './utils';
+import { StateStore } from './store';
+import type { Channel } from './channel';
+import type { StreamChat } from './client';
+import type {
+  ChannelFilters,
+  ChannelOptions,
+  ChannelSort,
+  DefaultGenerics,
+  ExtendableGenerics,
+  MessageFilters,
+  MessageResponse,
+  SearchMessageSort,
+  SearchOptions,
+  UserFilters,
+  UserOptions,
+  UserResponse,
+  UserSort,
+} from './types';
+
+export type SearchSourceType = 'channels' | 'users' | 'messages' | (string & {});
+export type QueryReturnValue<T> = { items: T[]; next?: string };
+export type DebounceOptions = {
+  debounceMs: number;
+};
+type DebouncedExecQueryFunction = DebouncedFunc<(searchString?: string) => Promise<void>>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export interface SearchSource<T = any> {
+  activate(): void;
+  deactivate(): void;
+  readonly hasNext: boolean;
+  readonly hasResults: boolean;
+  readonly initialState: SearchSourceState<T>;
+  readonly isActive: boolean;
+  readonly isLoading: boolean;
+  readonly items: T[] | undefined;
+  readonly lastQueryError: Error | undefined;
+  readonly next: string | undefined;
+  readonly offset: number | undefined;
+  resetState(): void;
+  search(text?: string): void;
+  searchDebounced: DebouncedExecQueryFunction;
+  readonly searchQuery: string;
+  setDebounceOptions(options: DebounceOptions): void;
+  readonly state: StateStore<SearchSourceState<T>>;
+  readonly type: SearchSourceType;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SearchSourceState<T = any> = {
+  hasNext: boolean;
+  isActive: boolean;
+  isLoading: boolean;
+  items: T[] | undefined;
+  searchQuery: string;
+  lastQueryError?: Error;
+  next?: string;
+  offset?: number;
+};
+
+export type SearchSourceOptions = {
+  /** The number of milliseconds to debounce the search query. The default interval is 300ms. */
+  debounceMs?: number;
+  isActive?: boolean;
+  pageSize?: number;
+};
+
+const DEFAULT_SEARCH_SOURCE_OPTIONS: Required<SearchSourceOptions> = {
+  debounceMs: 300,
+  isActive: false,
+  pageSize: 10,
+} as const;
+
+export abstract class BaseSearchSource<T> implements SearchSource<T> {
+  state: StateStore<SearchSourceState<T>>;
+  protected pageSize: number;
+  abstract readonly type: SearchSourceType;
+  searchDebounced!: DebouncedExecQueryFunction;
+
+  protected constructor(options?: SearchSourceOptions) {
+    const { debounceMs, isActive, pageSize } = { ...DEFAULT_SEARCH_SOURCE_OPTIONS, ...options };
+    this.pageSize = pageSize;
+    this.state = new StateStore<SearchSourceState<T>>({ ...this.initialState, isActive });
```

Comment:
I'd expect `this.initialState` to be constructed above:
```suggestion
    this.initialState = { 
        hasNext: true,
        isActive,
        isLoading: false,
        items: undefined,
        lastQueryError: undefined,
        next: undefined,
        offset: 0,
        searchQuery: '',
    }
    this.state = new StateStore<SearchSourceState<T>>(this.initialState);
```

Reply:
I have removed `isActive` from the constructor https://github.com/GetStream/stream-chat-js/pull/1434/commits/e145dae3dbb5d08f5152e4dae5eb3e61e033893d